worker/docker_client.py

```diff
@@ -0,0 +1,287 @@
+                os.path.join(cert_path, 'cert.pem'),
+                os.path.join(cert_path, 'key.pem'))
+            self._ssl_context.check_hostname = False
+
+        # Test to make sure that a connection can be established.
+        try:
+            self.test()
+        except DockerException:
+            print >> sys.stderr, """
+On Linux, a valid Docker installation should create a Unix socket at
+//var/run/docker.sock.
+
+On Mac, DOCKER_HOST and optionally DOCKER_CERT_PATH should be defined. You need
+to run the worker from the Docker shell.
+"""
+            raise
+        
+        # Find libcuda. We pick up the 64-bit version only and put it in a
+        # semi-standard directory. It works for the Tensorflow Docker image.
+        self._libcuda_files = []
+        try:
+            for lib in subprocess.check_output(['ldconfig', '-p']).split('\n'):
+                if 'libcuda.' not in lib or 'x86-64' not in lib:
+                    continue
+                self._libcuda_files.append(lib.split(' => ')[-1])
+        except OSError:
+            # ldconfig isn't available on Mac OS X. Let's just say that we
+            # don't support libcuda on Mac.
+            print >> sys.stderr, """
+No ldconfig found. Not loading libcuda libraries.
+"""
+
+        # Find all the NVIDIA device files.
+        self._nvidia_device_files = []
+        for filename in os.listdir('/dev'):
+            if filename.startswith('nvidia'):
+                self._nvidia_device_files.append(os.path.join('/dev', filename))
+
+    def _create_connection(self):
+        if self._docker_host:
+            if self._ssl_context:
+                return httplib.HTTPSConnection(self._docker_host,
+                                               context=self._ssl_context)
+            return httplib.HTTPConnection(self._docker_host)
+        return DockerUnixConnection()
+
+    @wrap_exception('Unable to connect to Docker')
+    def test(self):
+        with closing(self._create_connection()) as conn:
+            conn.request('GET', '/containers/json')
+            if conn.getresponse().status != 200:
+                raise DockerException(self._conn.getresponse().read())
+
+    @wrap_exception('Unable to download Docker image')
+    def download_image(self, docker_image, loop_callback):
+        logger.debug('Downloading Docker image %s', docker_image)
+        with closing(self._create_connection()) as conn:
+            conn.request('POST',
+                         '/images/create?fromImage=%s' % docker_image)
+            create_image_response = conn.getresponse()
+            if create_image_response.status != 200:
+                raise DockerException(create_image_response.read())
+
+            # Wait for the download to finish.
+            while True:
+                loop_callback()
+                response = None
+                line = ''
+                while True:
+                    ch = create_image_response.read(1)
```

Comment:
What's going on here?  This looks kind of weird/hacky - you have to keep on parsing something as JSON?  You can't just read the entire line and parse it as JSON?  Please document.